### ISSUE TITLE
Unpickling datetimes with timezones is not supported

### ISSUE THREAD

**Issue by airhorns** [irmen/Pyrolite#19]
At the moment Pyrolite doesn't support unpickling `datetime.date` or `datetime.datetime` objects  if they have a `tzinfo` property describing their timezone set. I would like to talk about adding support for this feature, but it's a bit tricky because of the different ways you can give a datetime a timezone.

For a datetime like this guy: `datetime.datetime(2014, 7, 8, 10)`, the `pickletools.dis` output looks like this:

```
    0: c    GLOBAL     'datetime datetime'
   19: p    PUT        0
   22: (    MARK
   23: S        STRING     '\x07\xde\x07\x08\n\x00\x00\x00\x00\x00'
   65: p        PUT        1
   68: t        TUPLE      (MARK at 22)
   69: p    PUT        2
   72: R    REDUCE
   73: p    PUT        3
   76: .    STOP
highest protocol among opcodes = 0
```

but for one like this:

```python
from dateutil.tz import tzutc
obj = datetime.datetime(2014, 7, 8, 10, 10, 0, 0, tzutc())
pickletools.dis(obj)
```

the pickle stream looks like this:

```
    0: c    GLOBAL     'datetime datetime'
   19: p    PUT        0
   22: (    MARK
   23: S        STRING     '\x07\xde\x07\x08\n\n\x00\x00\x00\x00'
   63: p        PUT        1
   66: c        GLOBAL     'copy_reg _reconstructor'
   91: p        PUT        2
   94: (        MARK
   95: c            GLOBAL     'dateutil.tz tzutc'
  114: p            PUT        3
  117: c            GLOBAL     'datetime tzinfo'
  134: p            PUT        4
  137: g            GET        4
  140: (            MARK
  141: t                TUPLE      (MARK at 140)
  142: R            REDUCE
  143: p            PUT        5
  146: t            TUPLE      (MARK at 94)
  147: p        PUT        6
  150: R        REDUCE
  151: p        PUT        7
  154: t        TUPLE      (MARK at 22)
  155: p    PUT        8
  158: R    REDUCE
  159: p    PUT        9
  162: .    STOP
```

which has that reconstructor call and then the tuple for the tzinfo instance which Pyrolite doesn't expect at all and fails to deserialize. Worse yet are tzinfos from pytz:

```python
from pytz import timezone
amsterdam = timezone('Europe/Amsterdam')
pickletools.dis(datetime.datetime(2014, 7, 8, 10, 10, 0, 0, amsterdam))
```

gives a pickle stream of 

```
    0: c    GLOBAL     'datetime datetime'
   19: p    PUT        0
   22: (    MARK
   23: S        STRING     '\x07\xde\x07\x08\n\n\x00\x00\x00\x00'
   63: p        PUT        1
   66: c        GLOBAL     'pytz _p'
   75: p        PUT        2
   78: (        MARK
   79: S            STRING     'Europe/Amsterdam'
   99: p            PUT        3
  102: I            INT        1200
  108: I            INT        0
  111: S            STRING     'LMT'
  118: p            PUT        4
  121: t            TUPLE      (MARK at 78)
  122: p        PUT        5
  125: R        REDUCE
  126: p        PUT        6
  129: t        TUPLE      (MARK at 22)
  130: p    PUT        7
  133: R    REDUCE
  134: p    PUT        8
  137: .    STOP
highest protocol among opcodes = 0
```

which has the `pytz._p` object in there with a string representing the zoneinfo name and the offset as well. You can also (and sadly I've seen code in the wild do this) use pytz's UTC representation like so:

```python
import pytz
utc = pytz.utc
pickletools.dis(datetime.datetime(2014, 7, 8, 10, 10, 0, 0, utc))
```

```
    0: c    GLOBAL     'datetime datetime'
   19: p    PUT        0
   22: (    MARK
   23: S        STRING     '\x07\xde\x07\x08\n\n\x00\x00\x00\x00'
   63: p        PUT        1
   66: c        GLOBAL     'pytz _UTC'
   77: p        PUT        2
   80: (        MARK
   81: t            TUPLE      (MARK at 80)
   82: R        REDUCE
   83: p        PUT        3
   86: t        TUPLE      (MARK at 22)
   87: p    PUT        4
   90: R    REDUCE
   91: p    PUT        5
   94: .    STOP
highest protocol among opcodes = 0
```

which is another somewhat annoying variation. Finally, I believe one could in a very twisted world give custom, extended objects as the `tzinfo` property that aren't of the base `dateutil.tz.tzinfo` class or `pytz._p` class. 

My suggestion/plan is to implement support for the second tuple element, and to hardcode an understanding of those two classes which almost everyone uses into the datetime unpickler, and still explode if someone passes a custom class that isn't registered in the java unpickler. If this approach sounds reasonable, I have some further questions:

 - Is there a better way to unpickle that second tuple element than the current approach where we bit shift out members of the byte string? Can I call out to a different piece of the system to get an object I can then inspect to build a java-side Timezone to add to the `Date` object?
 - In the first example above I see escaped newlines coming out in the pickle stream and I don't see that in any of the tests, any idea why my system might be doing that? 
 - Has anyone requested this functionality before or tried this kind of thing?

Thanks for any pointers or help or feedback you can give me!

I discovered this while debugging Apache Spark, so for posterity here's the Spark JIRA: https://issues.apache.org/jira/browse/SPARK-6411

**Comment by irmen** [comment]
Interesting issue. Are you willing to write a patch for it yourself? Note that the .net part of pyrolite will have to be fixed as well, probably by using DateTimeOffset.

We will likely have to register a few extra constructors in the Unpickler class to deal with the pytz/dateutil classes, to create a java Timezone object from them.  I suspect it should be possible to use the Unpickler itself to recursively reconstruct the timezone object from the pickle stream. 

I don't understand what you mean with "escaped newlines" and "any of the tests", can you explain this a bit better?

You're the first to hit this issue and the first to require timezone support in the datetime reconstructor :-)

**Comment by airhorns** [comment]
I do plan on writing a patch for the Java side :) However I have absolutely no experience with .net development or even a runtime to test stuff in so I would appreciate some help on that front. 

>  I suspect it should be possible to use the Unpickler itself to recursively reconstruct the timezone object from the pickle stream.

Cool, I think that makes sense. 

> I don't understand what you mean with "escaped newlines" and "any of the tests", can you explain this a bit better?

Yep, sorry. In `UnpicklerTests.java`, there are a couple tests asserting that de-pickling some given byte streams returns the expected java objects. The test here: https://github.com/irmen/Pyrolite/blob/cceabacacc5f2c095fe8a66d2215ce4ed2d8a72a/java/test/net/razorvine/pickle/test/UnpicklerTests.java#L235 depickles a byte string looking like this `"cdatetime\ndatetime\np0\n(S'\\x07\\xdb\\x0c\\x1f\\x0e!;\\x06\\xf5@'\np1\ntp2\nRp3\n."` which looks most similar to the first example I posted above where I'm just pickling a plain old datetime without a timezone, and works great. The `pickletools.dis` for this specific test byte string looks like this:

```
>>> bytes = "cdatetime\ndatetime\np0\n(S'\\x07\\xdb\\x0c\\x1f\\x0e!;\\x06\\xf5@'\np1\ntp2\nRp3\n."
>>> import pickle
>>> import pickletools
>>> pickle.loads(bytes)
datetime.datetime(2011, 12, 31, 14, 33, 59, 456000)
>>> pickletools.dis(bytes)
    0: c    GLOBAL     'datetime datetime'
   19: p    PUT        0
   22: (    MARK
   23: S        STRING     '\x07\xdb\x0c\x1f\x0e!;\x06\xf5@'
   58: p        PUT        1
   61: t        TUPLE      (MARK at 22)
   62: p    PUT        2
   65: R    REDUCE
   66: p    PUT        3
   69: .    STOP
highest protocol among opcodes = 0
```

which is all well and good, but then if I make myself a datetime with these particular values, I get a pickle byte stream with escaped newlines in it:

```
>>> y = datetime.datetime(2014, 7, 8, 10)
>>> pickletools.dis(pickle.dumps(y))
    0: c    GLOBAL     'datetime datetime'
   19: p    PUT        0
   22: (    MARK
   23: S        STRING     '\x07\xde\x07\x08\n\x00\x00\x00\x00\x00'
   65: p        PUT        1
   68: t        TUPLE      (MARK at 22)
   69: p    PUT        2
   72: R    REDUCE
   73: p    PUT        3
   76: .    STOP
highest protocol among opcodes = 0
```

In the STRING that starts at character 23 you'll see `\n`, and if I print out the raw bytestring, you can see that the newline in the middle of that string is escaped unlike the newline control chars:

```
>>> print pickle.dumps(y).encode('string-escape')
cdatetime\ndatetime\np0\n(S\'\\x07\\xde\\x07\\x08\\n\\x00\\x00\\x00\\x00\\x00\'\np1\ntp2\nRp3\n.
```

This pickle string fails to unpickle for me when I try to unpickle it in a Java land test with:

```
net.razorvine.pickle.PickleException: invalid escape sequence in string
	at net.razorvine.pickle.PickleUtils.decode_escaped(PickleUtils.java:335)
	at net.razorvine.pickle.Unpickler.load_string(Unpickler.java:441)
	at net.razorvine.pickle.Unpickler.dispatch(Unpickler.java:174)
	at net.razorvine.pickle.Unpickler.load(Unpickler.java:85)
	at net.razorvine.pickle.Unpickler.loads(Unpickler.java:98)
	at net.razorvine.pickle.test.UnpicklerTests.U(UnpicklerTests.java:48)
	at net.razorvine.pickle.test.UnpicklerTests.U(UnpicklerTests.java:43)
	at net.razorvine.pickle.test.UnpicklerTests.testDateTime(UnpicklerTests.java:235)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:601)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:50)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```

which doesn't seem quite right to me. I was wondering if A) I might be transporting the pickled bytes from my python shell to the Java test file wrong, or B) have I hit another unimplemented feature in Pyrolite where it is actually unable to deserialize escaped newlines, or secret option number C) that I don't see. Does that explain?

Thanks for any and all help!

**Comment by irmen** [comment]
I'm perfectly okay with creating the .net counterpart of a possible java patch myself, don't worry.

Regarding the newlines, this could indeed be a different bug. Glancing at the decode_escaped method I made it only look for "\\" or "\x" sequences which appears to be insufficient. Will create a separate issue for this.

**Comment by irmen** [comment]
Unfortunately this issue still has a problem.

I'm trying to call a Pyro object that's returning various datetime objects and while it seems to work with a pytz timezone, it still fails with a dateutil timezone.

The unpickling fails in the ClassDictConstructor with: expected zero arguments for construction of ClassDict (for dateutil.zoneinfo.gettz)

It seems there are still more classes to be added to the objectConstructors list of the Unpickler.

The python server code causing the problem is as follows:
```python
    def dateutil(self):
        tz_nl = dateutil.tz.gettz("Europe/Amsterdam")
        return datetime.datetime.now(tz_nl)
```

**Comment by irmen** [comment]
@airhorns I've added the problematic pickle as a (failing) unit test, unfortunately I have very little time this week to look at it myself.

**Comment by airhorns** [comment]
@irmen ok will take a look and have a patch for you shortly, sorry I missed that case!

**Comment by irmen** [comment]
No worries, and thanks for your efforts. Reading back your earlier notes it occurs to me only the tzutc was useful for your use case. I think the gettz("...") is another often occurring case.  For the other ways to get a dateutils tz, I don't know. Probably best to leave them be for now.

**Comment by airhorns** [comment]
@irmen what python version and  platform were you able to generate that pickle output on? I can only get stuff like this on Python 2.7.6 and 3.4.2 on OS X and Linux:

```
    0: \x80 PROTO      2
    2: c    GLOBAL     'datetime datetime'
   21: q    BINPUT     0
   23: U    SHORT_BINSTRING '\x07\xdf\x04\n\x15\x00,\x02}\xb3'
   35: q    BINPUT     1
   37: c    GLOBAL     'dateutil.tz tzfile'
   57: q    BINPUT     2
   59: U    SHORT_BINSTRING '/usr/share/zoneinfo/Europe/Amsterdam'
   97: q    BINPUT     3
   99: \x85 TUPLE1
  100: q    BINPUT     4
  102: R    REDUCE
  103: q    BINPUT     5
  105: \x86 TUPLE2
  106: q    BINPUT     6
  108: R    REDUCE
  109: q    BINPUT     7
  111: .    STOP
highest protocol among opcodes = 2
```

which is a constructor of a `tzfile` object instead of a call to the `gettz` function. Weird.

**Comment by airhorns** [comment]
Oh what version of dateutil would be good too,  I tried python-dateutil `2.2` and `2.4.2`

**Comment by irmen** [comment]
I created it on Python 2.7.8 on Windows 7, with dateutil version 2.4.2

It seems dateutil tz's pickle differently on different platforms. Yuck.

**Comment by irmen** [comment]
fixed by 4459b50fbc96cf1f0310202ad9581f331fe7175e
